pkg/volume/flexvolume/flexvolume.go

```diff
@@ -0,0 +1,310 @@
+	driverName string
+	// Driver executable used to setup the volume.
+	execPath string
+	// mounter provides the interface that is used to mount the actual
+	// block device.
+	mounter mount.Interface
+	plugin  *flexVolumePlugin
+}
+
+// FlexVolumeCleaner is the disk that will be cleaned by this plugin.
+type flexVolumeCleaner struct {
+	*flexVolumeDisk
+	// Runner used to teardown the volume.
+	runner exec.Interface
+	// manager is the utility interface that provides API calls to the
+	// driverName to setup & teardown disks
+	manager flexVolumeManager
+}
+
+// FlexVolumeBuilder is the disk that will be exposed by this plugin.
+type flexVolumeBuilder struct {
+	*flexVolumeDisk
+	// fsType is the type of the filesystem to create on the volume.
+	fsType string
+	// readOnly specifies whether the disk will be setup as read-only.
+	readOnly bool
+	// options are the extra params that will be passed to the plugin
+	// driverName.
+	options map[string]string
+	// Runner used to setup the volume.
+	runner exec.Interface
+	// manager is the utility interface that provides API calls to the
+	// driverName to setup & teardown disks
+	manager flexVolumeManager
+	// blockDeviceMounter provides the interface to create filesystem if the
+	// filesystem doesn't exist.
+	blockDeviceMounter mount.Interface
+}
+
+// SetUp creates new directory.
+func (f *flexVolumeBuilder) SetUp() error {
+	return f.SetUpAt(f.GetPath())
+}
+
+// flexVolumeManager is the abstract interface to flex volume ops.
+type flexVolumeManager interface {
+	// Attaches the disk to the kubelet's host machine.
+	setUp(builder *flexVolumeBuilder, dir string) (string, error)
+	// Detaches the disk to the kubelet's host machine.
+	teardown(cleaner *flexVolumeCleaner, dir, mntPath string) ([]byte, error)
+}
+
+// SetUpAt creates new directory.
+func (f *flexVolumeBuilder) SetUpAt(dir string) error {
+
+	notmnt, err := f.blockDeviceMounter.IsLikelyNotMountPoint(dir)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("Cannot validate mountpoint: %s", dir)
+		return err
+	}
+	if !notmnt {
+		return nil
+	}
+
+	device, err := f.manager.setUp(f, dir)
+	if err != nil {
+		glog.Errorf("Failed to setup volume: %s", f.volName)
+		return err
+	}
+
+	os.MkdirAll(dir, 0750)
+	var options []string
+
+	if f.readOnly {
+		options = []string{"ro"}
+	} else {
+		options = []string{"rw"}
+	}
+
+	err = f.blockDeviceMounter.Mount(string(device), dir, f.fsType, options)
+	if err != nil {
+		glog.Errorf("Failed to mount the volume: %s, device: %s, error: %s", f.volName, device, err.Error())
+	}
+	return err
+}
+
+// IsReadOnly returns true if the volume is read only.
+func (f *flexVolumeBuilder) IsReadOnly() bool {
+	return f.readOnly
+}
+
+// GetPathFromPlugin gets the actual volume mount directory based on plugin.
+func (f *flexVolumeDisk) GetPath() string {
+	name := f.driverName
+	return f.plugin.host.GetPodVolumeDir(f.podUID, util.EscapeQualifiedNameForDisk(name), f.volName)
```

Comment:
Interesting...  Driver name is of the format "Vendor/foo". We do not expect it to collide with existing plugins. Let me add a test case to verify the behavior if we collide.